ci/Jenkinsfile.test-e2e.android

```diff
@@ -0,0 +1,256 @@
+#!/usr/bin/env groovy
+library 'status-jenkins-lib@v1.9.27'
+
+pipeline {
+  agent {
+    dockerfile {
+      label 'linuxcontainer'
+      filename 'tests.Dockerfile'
+      dir 'ci'
+      args '--user jenkins'
+    }
+  }
+
+  parameters {
+    gitParameter(
+      name: 'GIT_REF',
+      description: 'Git branch to checkout.',
+      branchFilter: 'origin/(.*)',
+      branch: '',
+      defaultValue: 'master',
+      quickFilterEnabled: false,
+      selectedValue: 'DEFAULT',
+      sortMode: 'ASCENDING_SMART',
+      tagFilter: '*',
+      type: 'PT_BRANCH'
+    )
+    string(
+      name: 'BUILD_SOURCE',
+      description: 'URL to APK or Jenkins build (pkg/*.apk). Required when BROWSERSTACK_APP_ID is empty.',
+      defaultValue: ''
+    )
+    string(
+      name: 'BROWSERSTACK_APP_ID',
+      description: 'Existing BrowserStack app identifier (bs://...). Leave empty to upload from BUILD_SOURCE.',
+      defaultValue: ''
+    )
+    choice(
+      name: 'TEST_DEVICE_ID',
+      description: 'BrowserStack device id to run on (leave empty to use the environment default).',
+      choices: ['', 'galaxy_tab_s10p_android_15', 'pixel_8_android_14', 'pixel_7_android_13', 'samsung_s23_android_13']
+    )
+    string(
+      name: 'PYTEST_ARGS',
+      description: 'Pytest flags (e.g. "-m smoke" or "-m critical") and other args.',
+      defaultValue: '-n=5 -m smoke tests'
+    )
+  }
+
+  options {
+    timestamps()
+    timeout(time: 120, unit: 'MINUTES')
+    buildDiscarder(logRotator(
+      daysToKeepStr: '30',
+      numToKeepStr: '30',
+      artifactNumToKeepStr: '30',
+    ))
+    disableRestartFromStage()
+  }
+
+  environment {
+    VIRTUAL_ENV = "${env.WORKSPACE_TMP}/venv-appium"
+    PYTHONUNBUFFERED = "1"
+    TEST_DEVICE_ID = "${params.TEST_DEVICE_ID}"
+    BROWSERSTACK_PROJECT_NAME = "Mobile E2E ${utils.getBuildType()}"
+  }
+
+  stages {
+    stage('Prep') {
+      steps {
+        script {
+          setNewBuildName()
+          updateGitHubStatus()
+        }
+      }
+    }
+
+    stage('Setup Python environment') {
+      steps {
+        dir('test/e2e_appium') {
+          sh """
+            set -euo pipefail
+            python3 -m venv ${VIRTUAL_ENV}
+            source ${VIRTUAL_ENV}/bin/activate
+            pip install --upgrade pip
+            pip install -r requirements.txt
+          """
+        }
+      }
+    }
+
+    stage('Use provided BrowserStack app') {
+      when {
+        expression { params.BROWSERSTACK_APP_ID?.trim() }
+      }
+      steps {
+        script {
+          env.BROWSERSTACK_APP_ID = params.BROWSERSTACK_APP_ID.trim()
+          echo "Using provided BrowserStack app id: ${env.BROWSERSTACK_APP_ID}"
+        }
+      }
+    }
```

Comment:
Pointless stage that achieves nothing.

Reply:
Removed "used provided browserstack app" and "download" stages. When a URL for an APK is supplied (manual run) it is now passed to BrowserStack via the upload script to be downloaded directly.